concordium-smart-contract-testing/src/types.rs

```diff
@@ -0,0 +1,330 @@
+use std::{collections::BTreeMap, sync::Arc};
+use thiserror::Error;
+
+use concordium_base::{
+    base::Energy,
+    contracts_common::{
+        AccountAddress, AccountBalance, Address, Amount, ContractAddress, ExchangeRate,
+        ModuleReference, OwnedContractName, OwnedEntrypointName, SlotTime,
+    },
+    smart_contracts::WasmVersion,
+};
+use wasm_chain_integration::{
+    v0,
+    v1::{self, trie, ReturnValue},
+};
+use wasm_transform::artifact;
+
+/// A smart contract module.
+#[derive(Debug, Clone)]
+pub struct ContractModule {
+    /// Size of the module in bytes. Used for cost accounting.
+    pub(crate) size:     u64,
+    /// The runnable module.
+    pub(crate) artifact: Arc<artifact::Artifact<v1::ProcessedImports, artifact::CompiledFunction>>,
+}
+
+/// Represents the block chain and supports a number of operations, including
+/// creating accounts, deploying modules, initializing contract, updating
+/// contracts and invoking contracts.
+pub struct Chain {
+    /// The slot time viewable inside the smart contracts.
+    /// Defaults to `0`.
+    pub slot_time:           SlotTime,
+    /// MicroCCD per Euro ratio.
+    pub micro_ccd_per_euro:  ExchangeRate,
+    /// Euro per Energy ratio.
+    pub euro_per_energy:     ExchangeRate,
+    /// Accounts and info about them.
+    pub accounts:            BTreeMap<AccountAddress, Account>,
+    /// Smart contract modules.
+    pub modules:             BTreeMap<ModuleReference, ContractModule>,
+    /// Smart contract instances.
+    pub contracts:           BTreeMap<ContractAddress, Contract>,
+    /// Next contract index to use when creating a new instance.
+    pub next_contract_index: u64,
+}
+
+/// A smart contract instance.
+#[derive(Clone)]
+pub struct Contract {
+    /// The module which contains this contract.
+    pub module_reference: ModuleReference,
+    /// The name of the contract.
+    pub contract_name:    OwnedContractName,
+    /// The contract state.
+    pub state:            trie::PersistentState,
+    /// The owner of the contract.
+    pub owner:            AccountAddress,
+    /// The balance of the contract.
+    pub self_balance:     Amount,
+}
+
+/// Account policies for testing.
+pub struct TestPolicies(pub v0::OwnedPolicyBytes);
+
+/// An account.
+#[derive(Clone)]
+pub struct Account {
+    /// The account balance.
+    pub balance:         AccountBalance,
+    /// Account policies.
+    pub policies:        v0::OwnedPolicyBytes,
+    /// The number of signatures. The number of signatures affect the cost of
+    /// every transaction for the account.
+    pub signature_count: u32,
```

Comment:
I thought it would be easier to simply set it for an account instead of per transaction.
During development, I was unaware of your secret plans to eventually bring this into the node in some form. So I made choices based on what would be more convenient for the user, rather than always matching the node exactly 😊